packages/core-contracts/contracts/proxy/Beacon.sol

```diff
@@ -1,29 +1,32 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.0;
 
-import "./BeaconStorage.sol";
+import "./ProxyStorage.sol";
 import "../common/CommonErrors.sol";
-import "../interfaces/IBeacon.sol";
 import "../utils/ContractUtil.sol";
 
-contract Beacon is IBeacon, BeaconStorage, ContractUtil, CommonErrors {
-    event Upgraded(address indexed implementation);
+contract Beacon is ProxyStorage, ContractUtil, CommonErrors {
+    event Upgraded(address implementation);
 
-    function getImplementation() external view override returns (address) {
-        return _beaconStorage().implementation;
+    constructor(address firstImplementation) {
```

Comment:
Constructors are ok (and I think necessary) in proxies. It is the only way to set the first implementation without adding a an upgradeability admin selector to the ABI, which we can't have because of the proxy transparency problem.

Reply:
Hmmm or maybe not. Maybe the final rule could be that constructors can only be allowed in proxies, for simplicity of decisions.

Reply:
Ok, updated. Beacon.sol no longer has a constructor.